File: container-di/src/main/scala/com/yahoo/container/di/Osgi.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.di

import com.yahoo.config.FileReference
import com.yahoo.container.bundle.{MockBundle, BundleInstantiationSpecification}
import com.yahoo.container.di.Osgi.BundleClasses
import org.osgi.framework.Bundle
import com.yahoo.component.ComponentSpecification

/**
 *
 * @author gjoranv
 * @author tonytv
 */
trait Osgi {

  def getBundleClasses(bundle: ComponentSpecification, packagesToScan: Set[String]): BundleClasses = {
    BundleClasses(new MockBundle, List())
  }

  def useBundles(bundles: java.util.Collection[FileReference]) {
    println("useBundles " + bundles.toArray.mkString(", "))
  }

  def resolveClass(spec: BundleInstantiationSpecification): Class[AnyRef] = {
    println("resolving class " + spec.classId)
    Class.forName(spec.classId.getName).asInstanceOf[Class[AnyRef]]
  }

  def getBundle(spec: ComponentSpecification): Bundle = {
    println("resolving bundle " + spec)
    new MockBundle()
  }

}

object Osgi {
  type RelativePath = String //e.g. "com/yahoo/MyClass.class"
  case class BundleClasses(bundle: Bundle, classEntries: Iterable[RelativePath])
}

File: container-di/src/main/java/com/yahoo/container/di/Osgi.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.di;

import com.yahoo.component.ComponentSpecification;
import com.yahoo.config.FileReference;
import com.yahoo.container.bundle.BundleInstantiationSpecification;
import com.yahoo.container.bundle.MockBundle;
import com.yahoo.container.di.osgi.BundleClasses;
import org.osgi.framework.Bundle;

import java.util.Collection;
import java.util.Collections;
import java.util.Set;
import java.util.stream.Collectors;

import static java.util.Collections.emptySet;

/**
 * @author gjoranv
 * @author Tony Vaagenes
 * @author ollivir
 */
public interface Osgi {
    default BundleClasses getBundleClasses(ComponentSpecification bundle, Set<String> packagesToScan) {
        return new BundleClasses(new MockBundle(), Collections.emptySet());
    }

    default void installPlatformBundles(Collection<FileReference> bundles) {
        System.out.println("installPlatformBundles " + bundles.stream().map(Object::toString).collect(Collectors.joining(", ")));
    }

    /**
     * Returns the set of bundles that is not used by the current application generation,
     * and therefore should be scheduled for uninstalling.
     */
    default Set<Bundle> useBundles(Collection<FileReference> bundles) {
        System.out.println("useBundles " + bundles.stream().map(Object::toString).collect(Collectors.joining(", ")));
        return emptySet();
    }

    default Class<?> resolveClass(BundleInstantiationSpecification spec) {
        System.out.println("resolving class " + spec.classId);
        try {
            return Class.forName(spec.classId.getName());
        } catch (ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    default Bundle getBundle(ComponentSpecification spec) {
        System.out.println("resolving bundle " + spec);
        return new MockBundle();
    }
}
